File: 2024/albert-arvidsson-polyglot/day_03/part_2/solve.rb
#!/usr/bin/env ruby

regex = /(mul\((\d{1,3}),(\d{1,3})\)|do\(\)|don't\(\))/
enabled = true
sum = 0
File.read(ARGV[0]).scan(regex).each do |(match, one, two)|
  enabled = match.start_with? 'do(' if match.start_with? 'do'
  sum += one.to_i * two.to_i if enabled && match.start_with?('mul(')
end
pp sum

File: 2024/albert-arvidsson-polyglot/day_03/part_2/solve.mjs
#!/usr/bin/env node
import { readFileSync } from "node:fs";

const input = readFileSync(process.argv[2], "utf-8");
const regex = /mul\((\d{1,3}),(\d{1,3})\)|do\(\)|don't\(\)/g;
let sum = 0;
let enabled = true;
for (const [op, x, y] of input.matchAll(regex)) {
  if (op === "do()") enabled = true;
  else if (op === "don't()") enabled = false;
  else if (enabled) sum += +x * +y;
}
console.log(sum);
